lib/common/common/src/cpu.rs

```diff
@@ -0,0 +1,154 @@
+use std::cmp::Ordering;
+use std::sync::Arc;
+use std::thread;
+use std::time::Duration;
+
+use tokio::sync::{OwnedSemaphorePermit, Semaphore, TryAcquireError};
+
+use crate::defaults::default_cpu_budget_unallocated;
+
+/// Try to read number of CPUs from environment variable `QDRANT_NUM_CPUS`.
+/// If it is not set, use `num_cpus::get()`.
+pub fn get_num_cpus() -> usize {
+    match std::env::var("QDRANT_NUM_CPUS") {
+        Ok(val) => {
+            let num_cpus = val.parse::<usize>().unwrap_or(0);
+            if num_cpus > 0 {
+                num_cpus
+            } else {
+                num_cpus::get()
+            }
+        }
+        Err(_) => num_cpus::get(),
+    }
+}
+
+/// Get available CPU budget to use for optimizations as number of CPUs (threads).
+///
+/// This is user configurable via `cpu_budget` parameter in settings:
+/// If 0 - auto selection, keep at least one CPU free when possible.
+/// If negative - subtract this number of CPUs from the available CPUs.
+/// If positive - use this exact number of CPUs.
+///
+/// The returned value will always be at least 1.
+pub fn get_cpu_budget(cpu_budget_param: isize) -> usize {
+    match cpu_budget_param.cmp(&0) {
+        // If less than zero, subtract from available CPUs
+        Ordering::Less => get_num_cpus()
+            .saturating_sub(-cpu_budget_param as usize)
+            .max(1),
+        // If zero, use automatic selection
+        Ordering::Equal => get_num_cpus()
+            .saturating_sub(-default_cpu_budget_unallocated(get_num_cpus()) as usize)
+            .max(1),
+        // If greater than zero, use exact number
+        Ordering::Greater => cpu_budget_param as usize,
+    }
+}
+
+/// Structure managing global CPU budget for optimization tasks.
+///
+/// Assigns CPU permits to tasks to limit overall resource utilization, making optimization
+/// workloads more predictable and efficient.
+#[derive(Debug, Clone)]
+pub struct CpuBudget {
+    semaphore: Arc<Semaphore>,
+}
+
+impl CpuBudget {
+    pub fn new(cpu_budget: usize) -> Self {
+        Self {
+            semaphore: Arc::new(Semaphore::new(cpu_budget)),
+        }
+    }
+
+    /// Try to acquire CPU permit for optimization task from global CPU budget.
+    pub fn try_acquire(&self, desired_cpus: usize) -> Option<CpuPermit> {
+        // Determine what number of CPUs to acquire based on available budget
+        let num_cpus = self.semaphore.available_permits().min(desired_cpus) as u32;
```

Comment:
I don't think we should allow creation of permits with less then `desired_cpus` number of cores (except for the cases, where total number of CPUs are less than `desired_cpus`).


Imagine the situation, where we have queue of 2 tasks with `desired_cpus = 7`  and 8 CPUs available. 

That will generate permit for 7 and 1 CPUs respectively. Assuming linear CPU/task efficiency, the total time of the optimization will be defined by the slowest task - the one which was executed by 1 CPU. Let's say it is X seconds.

The other task will be completed in X/7 seconds, which will left 7 CPUs un-utilized for X * 6/7 seconds.


A scenario where we schedule tasks only if `desired_cpus` is available would take X * 2/7 seconds

Reply:
But I would agree, if we make the limit to be `self.semaphore.available_permits() > (desired_cpus) / 2 + 1`

Reply:
The new implementation uses the _'at least half'_ approach: `semaphore.available_permits() > ceil(desired_cpus / 2)`.

It also limits the desired CPU to the total CPU budget to prevent having abnormal 'minimum required' numbers.